Add utility script to collect metrics data

reidmv opened the pull request:
The purpose of this utility is to enable assist and support operations.
This commit adds a script which streamlines the user experience of
collecting and submitting metrics data for analysis.

Fixes #40

reidmv committed:
Add utility script to collect metrics data

The purpose of this utility is to enable assist and support operations.
This commit adds a script which streamlines the user experience of
collecting and submitting metrics data for analysis.

Fixes #40

npwalker commented on manifests/init.pp:
```diff
@@ -31,6 +32,14 @@
     source  => 'puppet:///modules/pe_metric_curl_cron_jobs/tk_metrics'
   }
 
+  file { '/opt/puppetlabs/bin/puppet-metrics-collector':
```
`/opt/puppetlabs/bin` doesn't appear to be in the $PATH by default.  Is it beneficial to put this there?  

```
[root@master201646-centos ~]# ls -l /opt/puppetlabs/bin/
total 144
lrwxrwxrwx 1 root root     52 Jul 21 23:55 console-services -> ../server/apps/console-services/bin/console-services
lrwxrwxrwx 1 root root     33 Jul 21 23:53 facter -> /opt/puppetlabs/puppet/bin/facter
lrwxrwxrwx 1 root root     32 Jul 21 23:53 hiera -> /opt/puppetlabs/puppet/bin/hiera
lrwxrwxrwx 1 root root     30 Jul 21 23:53 mco -> /opt/puppetlabs/puppet/bin/mco
lrwxrwxrwx 1 root root     64 Jul 21 23:55 orchestration-services -> ../server/apps/orchestration-services/bin/orchestration-services
lrwxrwxrwx 1 root root     33 Jul 21 23:53 pe-man -> /opt/puppetlabs/server/bin/pe-man
lrwxrwxrwx 1 root root     33 Jul 21 23:53 puppet -> /opt/puppetlabs/puppet/bin/puppet
lrwxrwxrwx 1 root root     46 Jul 21 23:53 puppet-access -> /opt/puppetlabs/client-tools/bin/puppet-access
lrwxrwxrwx 1 root root     43 Jul 21 23:53 puppet-app -> /opt/puppetlabs/client-tools/bin/puppet-app
lrwxrwxrwx 1 root root     44 Jul 21 23:53 puppet-code -> /opt/puppetlabs/client-tools/bin/puppet-code
lrwxrwxrwx 1 root root     36 Jul 21 23:55 puppetdb -> ../server/apps/puppetdb/bin/puppetdb
lrwxrwxrwx 1 root root     42 Jul 21 23:53 puppet-db -> /opt/puppetlabs/client-tools/bin/puppet-db
lrwxrwxrwx 1 root root     60 Jul 21 23:55 puppet-enterprise -> /opt/puppetlabs/server/apps/enterprise/bin/puppet-enterprise
-rwxr-xr-x 1 root root 136894 Jul 21 23:53 puppet-enterprise-uninstaller
lrwxrwxrwx 1 root root     43 Jul 21 23:53 puppet-job -> /opt/puppetlabs/client-tools/bin/puppet-job
lrwxrwxrwx 1 root root     45 Jul 21 23:53 puppet-query -> /opt/puppetlabs/client-tools/bin/puppet-query
lrwxrwxrwx 1 root root     44 Jul 21 23:53 puppetserver -> ../server/apps/puppetserver/bin/puppetserver
lrwxrwxrwx 1 root root     31 Jul 21 23:54 r10k -> /opt/puppetlabs/puppet/bin/r10k
[root@master201646-centos ~]# echo $PATH
/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin:/opt/puppetlabs/bin:/root/bin
[root@master201646-centos ~]# ls -l /usr/local/bin/
total 0
lrwxrwxrwx 1 root root 26 Jul 21 23:54 facter -> /opt/puppetlabs/bin/facter
lrwxrwxrwx 1 root root 25 Jul 21 23:54 hiera -> /opt/puppetlabs/bin/hiera
lrwxrwxrwx 1 root root 23 Jul 21 23:54 mco -> /opt/puppetlabs/bin/mco
lrwxrwxrwx 1 root root 26 Jul 21 23:54 pe-man -> /opt/puppetlabs/bin/pe-man
lrwxrwxrwx 1 root root 26 Jul 21 23:54 puppet -> /opt/puppetlabs/bin/puppet
lrwxrwxrwx 1 root root 24 Jul 21 23:54 r10k -> /opt/puppetlabs/bin/r10k
```

reidmv replied:
I like the idea of putting scripts intended to be executed by users in a `bin/` directory. We don't have one in the module. `/opt/puppetlabs/bin` is already a known location we put our utilities in, and contains many more utilities than the ones we link to `/usr/local/bin`. It seems to be the appropriate place for it. The possible counter-argument is for self-containment, in which case creating a `/opt/puppetlabs/pe_metric_curl_cron_jobs/bin` directory would be the equivalent.

We currently have a per-component bin dir already in our `/opt/puppetlabs/<thing>` pattern, but we symlink back everything we expect users to run to `/opt/puppetlabs/bin`.

reidmv committed:
Implement bin_dir / symlink pattern

To mimic existing Puppet utilities and bindir behavior. Create a local
bin dir, but provide a symlink to user-facing utilities in
/opt/puppetlabs/bin

reidmv commented:
Updated PR to implement local bin dir, symlink in `/opt/puppetlabs/bin`

npwalker commented:
@Sharpie @jarretlavallee thoughts / concerns?

jarretlavallee commented:
👍 Looks good to me.

npwalker commented:
@reidmv can you add something to the README?  We'll need to update the knowledge base article as well if we want people to actually know it's available to use.

reidmv committed:
Update utility script to support $output_dir

Previously the value of $output_dir was hard-coded and assumed in the
utility script. This commit updates the script to be built by template
and use the value of $output_dir, whatever it may be.

reidmv commented:
@npwalker README section added, also updated the utility script to honor `$output_dir` (forgot about that on the first try)

reidmv commented:
@npwalker merge?

npwalker committed:
Change puppet-metrics-collector to always be present

and optionally symlinked.

npwalker commented:
@reidmv I modified it with another commit, let me know if you're good with it

npwalker commented:
If we someday rename the module to puppet_metrics_collector will having this script named that be problematic?

reidmv commented:
@npwalker looks good to me!

reidmv commented:
@npwalker naming the script puppet-metrics-collector is in fact intentional, and for precisely the reason that I anticipate this repo being named similar. That's also why the invocation requires a subcommand; to leave room for other CLI actions besides 'create-tarball'.
